README
This is a library that implements encoding and parsing of DNS queries
and responses, per RFC 1035.

Documentation of library functions is found in rfc1035.h, along with
the declarations of those functions.  They are implemented in rfc1035.c

testdns.c is a simple program for doing a query and looking at the result.

check_dnsauth.c is a more complex example; it acts as a Nagios check
and will query a set of nameservers and verify their consistency.

See TODO for caveats.

This is GPL'ed.

Regards,
adb


ChangeLog
Oct 25 2006 - release 0.3

 - considered LGPL, decided against it (sorry PdaS)
 - fixed fcntl() calling (moment of silence for WRS)
 - fixed bug with query IDs in check_dnsauth (thanks Shayne for spotting it)

Nov 21 2005 - release 0.2

check_dnsauth fixes
 - use answer section of response only
 - report ERROR or NXDOMAIN only once
 - don't include getopt.h or clear optind
 - fix thinko in location of socket open

added ChangeLog

Nov 20 2005 - release 0.1


rfc1035.c
/* rfc1035.c

  Copyright (C) 2005 Anthony de Boer

  This program is free software; you can redistribute it and/or modify
  it under the terms of version 2 of the GNU General Public License as
  published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

#include "rfc1035.h"

int dnsencode(char *name, char *buf, int id, int rd, int qtype, int qclass) {
	char *q, *p, c;
	memset(buf+3, 0, 10);
	buf[0] = (id>>8) & 0377;
	buf[1] = id & 0377;
	buf[2] = rd ? 1 : 0;
	buf[5] = 1;
	q = buf+12;
	p = buf+13;
	strcpy(p, name);
	while(c = (*p++)) {
		if (c == '.') {
			q = p - 1;
			*q = 0;
			}
		else {
			(*q)++;
			}
		}
	*p++ = (qtype >> 8) & 0377;
	*p++ = qtype & 0377;
	*p++ = (qclass >> 8) & 0377;
	*p++ = qclass & 0377;
	return p - buf;
	}

unsigned int dnsget16(unsigned char *p) {
	unsigned int rc = (p[0] & 0377) << 8;
	rc |= (p[1] & 0377);
	return rc;
	}

unsigned int dnsget32(unsigned char *p) {
	unsigned int rc = (p[0] & 0377) << 24;
	rc |= (p[1] & 0377) << 16;
	rc |= (p[2] & 0377) << 8;
	rc |= (p[3] & 0377);
	return rc;
	}

int dnsgetstr(unsigned char *buf, int blen, int noff, char *sb, int sl) {
	int nl;
	int rc = -1;
	int nr = 0;
	while (1) {
		if (noff >= blen) goto toolong;
		nl = (0377 & buf[noff++]);
		if ((nl & 0300) == 0300) {
			if (noff >= blen) goto toolong;
			noff = ((nl & 077) << 8) | (0377 & buf[noff]);
			}
		else if (nl) {
			int i;
			if (noff + nl > blen) goto toolong;
			if (nr + nl >= sl) goto toolong;
			memcpy(sb+nr, buf+noff, nl);
			nr += nl;
			noff += nl;
			sb[nr++] = '.';
			}
		else {
			break;
			}
		}
	if (nr>1) {
		nr--;
		}
	sb[nr] = '\0';
	rc = nr;
	toolong:
	return rc;
	}

void dnsparse(unsigned char *buf, int blen, dnscbf dcfun, void *edata) {
	unsigned int nrx[4];
	unsigned int rtype, rclass, rttl, rsize;
	int rind = 0;
	int j = 12;
	int nameoff;

	nrx[0] = dnsgetnqueries(buf);
	nrx[1] = dnsgetnanswers(buf);
	nrx[2] = dnsgetnauth(buf);
	nrx[3] = dnsgetnadd(buf);

	while (j < blen) {
		while (nrx[rind] == 0) {
			rind++;
			if (rind > 3) goto consideredharmful;
			}
		nrx[rind]--;
		nameoff = j;
		while (1) {
			int mbyte = buf[j++];
			if ((mbyte & 0300) == 0300) {
				j++;
				break;
				}
			else if (mbyte) {
				j += mbyte;
				if (j >= blen) goto consideredharmful;
				}
			else {
				break;
				}
			}
		if (j+4 > blen) goto consideredharmful;
		rtype = dnsget16(buf+j);
		rclass = dnsget16(buf+j+2);
		j += 4;
		if (rind > 0) {
			if (j+6 > blen) goto consideredharmful;
			rttl = dnsget32(buf+j);
			rsize = dnsget16(buf+j+4);
			j += 6;
			if (j+rsize > blen) goto consideredharmful;
			}
		else {
			rttl = rsize = 0;
			}
		dcfun(buf, blen, rind, nameoff, rtype, rclass, rttl, rsize, j, edata);
		j += rsize;
		}
	consideredharmful: {}
	}

char *qtypes[] = {
	0, "A", "NS", "MD", "MF", "CNAME", "SOA", "MB",
	"MG", "MR", "NULL", "WKS", "PTR", "HINFO", "MINFO", "MX",
	"TXT", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, "AXFR", "MAILB", "MAILA", "ANY",
	};

char *dnsqtypename(int qtype) {
	return (qtype < 0 || qtype > (sizeof(qtypes)/sizeof(char*))) ? 0 : qtypes[qtype];
	}

int dnsrrstr(unsigned char *buf, int blen, unsigned int rtype, unsigned int rclass, unsigned int rsize, unsigned int recoff, char *sb, int sl) {
	int rc;
	char *qts = dnsqtypename(rtype);
	if (!qts || rclass != 1 || !sl) {
		rc = -1;
		}
	else if (rsize == 0) {
		rc = snprintf(sb, sl, "(no RR)");
		}
	else if (rtype == 1 && rsize == 4) {
		/* IP addr */
		rc = snprintf(sb, sl, "A %u.%u.%u.%u", buf[recoff], buf[recoff+1], buf[recoff+2], buf[recoff+3]);
		}
	else if (rtype == 2 || rtype == 5 || rtype == 12) {
		rc = snprintf(sb, sl, "%s ", qts);
		if (rc < 0 || rc >= sl) {
			goto rroops;
			}
		else {
			sb += rc;
			sl -= rc;
			}
		rc = dnsgetstr(buf, blen, recoff, sb, sl);
		}
	else if (rtype == 15 && rsize > 2) {
		unsigned int mxpref;
		mxpref = dnsget16(buf+recoff);
		rc = snprintf(sb, sl, "%s %u ", qts, mxpref);
		if (rc < 0 || rc >= sl) {
			goto rroops;
			}
		else {
			sb += rc;
			sl -= rc;
			}
		rc = dnsgetstr(buf, blen, recoff+2, sb, sl);
		}
	else {
		rc = -1;
		}
	rroops: return (rc > 0 && rc < sl) ? 0 : -1;
	}


rfc1035.h
/* rfc1035.h

  Copyright (C) 2005 Anthony de Boer

  This program is free software; you can redistribute it and/or modify
  it under the terms of version 2 of the GNU General Public License as
  published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

/* --- encoding routine --- */

/* dnsencode: encode a DNS query.  Passed:
 name - string name to query (eg. "www.example.com")
 buf - a buffer in which to store the query; must be at least 17 octets
 longer than the name.
 id - query ID, 16-bit quantity
 rd - nonzero for recursion-desired, zero for no recursion
 qtype - query type (1==A)
 qclass - query class (1=IN)
Returns number of octets encoded in buf.
*/
int dnsencode(char *name, char *buf, int id, int rd, int qtype, int qclass);

/* --- decoding routines --- */

/* for internal use only */
unsigned int dnsget16(unsigned char *p);

/* return query ID from response buffer x */
#define dnsgetid(x) dnsget16(x)

/* return flags word from response buffer x */
#define dnsgetflags(x) dnsget16((x)+2)

/* return the number of contained queries from response buffer x */
#define dnsgetnqueries(x) dnsget16((x)+4)

/* return the number of contained answers from response buffer x */
#define dnsgetnanswers(x) dnsget16((x)+6)

/* return the number of contained authority records from response buffer x */
#define dnsgetnauth(x) dnsget16((x)+8)

/* return the number of contained additional records from response buffer x */
#define dnsgetnadd(x) dnsget16((x)+10)

/* return the authority flag from response buffer x */
#define dnsgetaa(x) ((dnsgetflags(x) >> 10) & 1)

/* return the response code from response buffer x (0=OK, 3=NXDOMAIN,
other nonzero values indicate other problems per 1035 page 27 */
#define dnsgetrcode(x) (dnsgetflags(x) & 15)

/* Callback that gets asked about each parsed record; you have to write one.
 buf - the original buffer
 blen - size thereof
 rind - type of record (0=question, 1=answer, 2=authority, 3=additional)
 nameoff - offset in buf at which name starts
 rtype - record type (1=A etc)
 rclass - record class (1=IN)
 rttl - TTL
 rsize - size of resource data
 recoff - offset in buf at which resource data starts
 edata - extra data (cast a pointer to your own struct, whatever)
*/
typedef void (*dnscbf)(unsigned char *buf, int blen, int rind, int nameoff, unsigned int rtype, unsigned int rclass, unsigned int rttl, unsigned int rsize, unsigned int recoff, void *edata);

/* parser:
 buf - buffer containing DNS packet
 blen - length of packet
 dcfun - callback function that will be called for each record (see above)
 edata - extra data pointer to pass to callback (so you can tell it which
 record it's in at the time)
*/
void dnsparse(unsigned char *buf, int blen, dnscbf dcfun, void *edata);

/* get a labelled string out of a DNS record.  This is useful in your
callback routine to either get the name (using nameoff, in all cases) or
using recoff if the payload is a name.
 buf - the original buffer
 blen - size thereof
 noff - name offset in record
 sb - string buffer
 sl - size of buffer
 RETURN: length of string, or -1 on error
*/
int dnsgetstr(unsigned char *buf, int blen, int noff, char *sb, int sl);

#define DNSQTYPEMIN 1
#define DNSQTYPEMAX 255

/* return the name of a query type.  NULL is returned for invalid ones.
 1 is "A" and so on.  The useful range is DNSQTYPEMIN through DNSQTYPEMAX.
*/
char *dnsqtypename(int qtype);

/* generate a string describing an RR.  Pass:
 buf - the original buffer
 blen - size thereof
 rtype - record type (1=A etc)
 rclass - record class (1=IN)
 rsize - size of resource data
 recoff - offset in buf at which resource data starts
 sb - string buffer
 sl - size of buffer
Returns 0 on success.
*/
int dnsrrstr(unsigned char *buf, int blen, unsigned int rtype, unsigned int rclass, unsigned int rsize, unsigned int recoff, char *sb, int sl);


testdns.c
/* testdns.c

  Copyright (C) 2005 Anthony de Boer

  This program is free software; you can redistribute it and/or modify
  it under the terms of version 2 of the GNU General Public License as
  published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

#include <stdio.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>

#include "rfc1035.h"

int dnsfd;

void sloop();
void my_log_data(char *mydata, int mylen);

void barf(char *s) {
	perror(s);
	exit(1);
	}

void usage() {
	fprintf(stderr, "abusage!\n");
	exit(3);
	}

int main(int argc, char **argv) {
	char buf[512];
	struct sockaddr_in sadr;
	int rlen, rc;
	char *who = NULL;
	char *s;
	int qtype = 1;
	int authq = 0;
	int prand;
	int i;
	prand = getpid();

	while(1) {
		switch(getopt(argc, argv, "w:At:")) {
			case 'w':
				who = strdup(optarg);
				break;
			case 'A':
				authq = 1;
				break;
			case 't':
				qtype = atoi(optarg);
				for (i=DNSQTYPEMIN;i<=DNSQTYPEMAX;i++) {
					s = dnsqtypename(i);
					if (s && !strcmp(optarg, s)) {
						qtype = i;
						break;
						}
					}
				if (!qtype) {
					fprintf(stderr, "Bad -t option\n");
					exit(3);
					}
				break;
			case EOF:
				goto doneopts;
			default:
				usage();
			}
		}
	doneopts:

	if (!who) {
		usage();
		}

	rlen = dnsencode(who, buf, prand++, authq?0:1, qtype, 1);
	my_log_data (buf, rlen);

	dnsfd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
	if (dnsfd == -1) barf("socket");

	bzero(&sadr, sizeof(sadr));
	sadr.sin_family = AF_INET;
	sadr.sin_port = htons(53);
	sadr.sin_addr.s_addr = inet_addr("192.168.240.1");
	rc = sendto(dnsfd, buf, rlen, 0, (struct sockaddr *)&sadr, sizeof(sadr));
	if (rc == -1) barf("sendto");

	sloop();
	}

void dnsrec(unsigned char *buf, int blen, int rind, int nameoff, unsigned int rtype, unsigned int rclass, unsigned int rttl, unsigned int rsize, unsigned int recoff, void *edata) {
	char *aty[] = { "Query", "Answer", "Authority", "Additional", };
	char sbuf[256];
	int rc;
	rc = dnsgetstr(buf, blen, nameoff, sbuf, sizeof(sbuf));
	printf("%s", rc < 0 ? "ERROR" : sbuf);
	rc = dnsrrstr(buf, blen, rtype, rclass, rsize, recoff, sbuf, sizeof(sbuf));
	printf(" - %s %s\n", aty[rind], rc ? "ERROR" : sbuf);
	}

/* Called when data is ready to read on socket s */
void my_read() {
	int rc;
	char buf[1024];
	rc = read(dnsfd, buf, sizeof(buf));
	if (rc == -1) barf("read");
	write(2, "\ngot\n", 5);
	my_log_data(buf, rc);
	fprintf(stderr, "parsing...\n");
	dnsparse(buf, rc, dnsrec, NULL);
	fprintf(stderr, "parsed...\n");
	close(dnsfd);
	}

void sloop() {
	int rc;
	fd_set readfds, writefds, exceptfds;
	struct timeval timeout;
	while(1) {
		FD_ZERO(&readfds);
		FD_ZERO(&writefds);
		FD_ZERO(&exceptfds);
		FD_SET(dnsfd, &readfds);
		FD_SET(dnsfd, &exceptfds);
		timeout.tv_sec = 5;
		timeout.tv_usec = 0;
		fprintf(stderr, "selecting...\n");
		rc = select(dnsfd+1, &readfds, &writefds, &exceptfds, &timeout);
		if (rc > 0) {
			fprintf(stderr, "something...\n");
			if (FD_ISSET(dnsfd, &readfds)) {
				my_read();
				}
			else if (FD_ISSET(dnsfd, &exceptfds)) {
				fprintf(stderr, "Exception on DNS socket\n");
				exit(3);
				}
			}
		else if (rc < 0 && errno != EINTR) {
			perror("select()");
			exit(3);
			}
		}
	}

#define hexdig(q) (hdigs[(q) & 0xf])

void my_log_data(char *mydata, int mylen) {
	char obuf[65];
	char *hdigs = "0123456789ABCDEF";
	char *oo1, *oo2;
	int i, v, x, z;
	while(mylen > 0) {
		oo1 = obuf;
		oo2 = obuf+48;
		z = mylen < 16 ? mylen : 16;
		for (i=0; i<16; i++) {
			if (v = (i<mylen))
				x = *(mydata++);
			*(oo1++) = v ? hexdig(x>>4) : '_';
			*(oo1++) = v ? hexdig(x) : '_';
			*(oo1++) = ' ';
			*(oo2++) = v ? (x < ' ' || x > '~') ? '.' : x : '\0';
			}
		obuf[48+z] = '\n';
		write(2, obuf, z+49);
		mylen -= 16;
		}
	}


check_dnsauth.c
/* check_dnsauth.c

  Copyright (C) 2005,2006 Anthony de Boer

  This program is free software; you can redistribute it and/or modify
  it under the terms of version 2 of the GNU General Public License as
  published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

#include <stdio.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <arpa/inet.h>

#include "rfc1035.h"

#define WARNBIT 1
#define CRITBIT 2

struct nsnode {
	char *nsname;
	char *nsip;
	int queryno;
	int npos;
	int resprec;
	int critns;
	struct nsnode *next;
	};

struct rrnode {
	char *rrname;
	int bits;
	struct rrnode *next;
	};

/* glowball */
int dnsfd;
struct nsnode *nshead = NULL;
struct rrnode *rrhead = NULL;
int nbit = 1;
int expbit = 0;
char *who = NULL;
int authq = 0;
int qtype = 1;
int debug = 0;
int syncsev = WARNBIT;
int prand;

void sloop();

void barf(char *s) {
	perror(s);
	exit(3);
	}

void puke(char *s) {
	fprintf(stderr, "%s\n", s);
	exit(3);
	}

void upchuck(char *s, char *t) {
	fprintf(stderr, "%s: \"%s\"\n", s);
	exit(3);
	}

void newns(char *name, int crit) {
	struct nsnode *tns;
	char *scp, *p;
	tns = (struct nsnode *)malloc(sizeof(struct nsnode));
	if (!tns) puke("malloc failure");
	scp = strdup(name);
	p = index(scp, ':');
	if (!p) upchuck("-n arg without a colon", scp);
	*p++ = '\0';
	tns->nsname = scp;
	tns->nsip = p;
	tns->next = nshead;
	tns->queryno = prand++;
	tns->npos = nbit;
	tns->critns = crit ? CRITBIT : WARNBIT;
	tns->resprec = strcmp(scp, "_expect") ? 0 : 1;
	nbit <<= 1;
	nshead = tns;
	}

struct rrnode *findrr(char *name) {
	struct rrnode *prr;
	prr = rrhead;
	while(prr) {
		if (!strcmp(name, prr->rrname)) goto foundrr;
		prr = prr->next;
		}
	prr = (struct rrnode *)malloc(sizeof(struct rrnode));
	if (!prr) puke("malloc failure");
	prr->rrname = strdup(name);
	if (!prr->rrname) puke("malloc failure");
	prr->bits = 0;
	prr->next = rrhead;
	rrhead = prr;
	foundrr:
	return prr;
	}

void newexp(char *name) {
	struct rrnode *prr = findrr(name);
	if (expbit == 0) {
		expbit = nbit;
		newns("_expect:0.0.0.0", 0);
		}
	prr->bits |= expbit;
	}

void endgame() {
	struct rrnode *prr = rrhead;
	struct nsnode *nrr = nshead;
	int allbits = nbit - 1;
	int severity = 0;
	int fudge = 0;
	char *sep = " -";
	while(prr) {
		if (debug) printf("%s has %d\n", prr->rrname, prr->bits);
		if (prr->bits != allbits) severity |= syncsev;
		prr = prr->next;
		}
	while(nrr) {
		if (nrr->resprec != 1) severity |= nrr->critns;
		nrr = nrr->next;
		}
	printf("%s %s", who, (severity & CRITBIT) ? "CRITICAL" : severity ? "WARNING" : "OK");
	nrr = nshead;
	while(nrr) {
		char *prob;
		switch (nrr->resprec) {
			case 0: prob = "NORESPONSE"; break;
			case -1: prob = "ERROR"; break;
			case -3: prob = "NXDOMAIN"; break;
			case 1: prob = NULL; break;
			default: prob = "WTF?";
			}
		if (prob) {
			printf("%s %s %s", sep, nrr->nsname, prob);
			sep = ",";
			fudge |= nrr->npos;
			}
		nrr = nrr->next;
		}
	prr = rrhead;
	while(prr) {
		printf("%s %s", sep, prr->rrname);
		sep = ",";
		if ((prr->bits | fudge) != allbits) {
			printf(" {");
			nrr = nshead;
			while(nrr) {
				if (nrr->npos & prr->bits) {
					printf(" %s", nrr->nsname);
					}
				nrr = nrr->next;
				}
			printf(" NOT");
			nrr = nshead;
			while(nrr) {
				if (nrr->resprec == 1 && !(nrr->npos & prr->bits)) {
					printf(" %s", nrr->nsname);
					}
				nrr = nrr->next;
				}
			printf("}");
			}
		prr = prr->next;
		}
	printf ("\n");
	exit ((severity & CRITBIT) ? 2 : severity ? 1 : 0);
	}

void usage() {
	fprintf(stderr, "Usage:\n"
	"  -w target\n"
	"  -A (auth query only)\n"
	"  -t TYPE\n"
	"  -n nsname:ipno (WARNING on fail)\n"
	"  -N nsname:ipno (CRITICAL on fail)\n"
	"  -e expectation\n"
	"  -C (critical on sync fail)\n"
	"  -d (debug)\n"
	);
	exit(3);
	}

int main(int argc, char **argv) {
	char *s;
	int i;
	int flags;

	prand = getpid();

	while(1) {
		switch(getopt(argc, argv, "w:At:n:N:Ce:d")) {
			case 'w':
				who = strdup(optarg);
				break;
			case 'A':
				authq = 1;
				break;
			case 't':
				qtype = atoi(optarg);
				for (i=DNSQTYPEMIN;i<=DNSQTYPEMAX;i++) {
					s = dnsqtypename(i);
					if (s && !strcmp(optarg, s)) {
						qtype = i;
						break;
						}
					}
				if (!qtype) puke("Bad -t option");
					
				break;
			case 'n':
				newns(optarg, 0);
				break;
			case 'N':
				newns(optarg, 1);
				break;
			case 'C':
				syncsev = CRITBIT;
				break;
			case 'e':
				newexp(optarg);
				break;
			case 'd':
				debug = 1;
				break;
			case EOF:
				goto doneopts;
			default:
				usage();
			}
		}
	doneopts:

	if (!who) usage();

	dnsfd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
	if (dnsfd == -1) barf("socket");

	/* see UNPv2 p58 */
	if ((flags = fcntl(dnsfd, F_GETFL, 0)) == -1) barf("fcntl F_GETFL");
	flags |= O_NONBLOCK;
	if (fcntl(dnsfd, F_SETFL, flags) == -1) barf("fcntl F_SETFL");

	sloop();
	endgame();
	}

int sendqueries() {
	char buf[512];
	struct sockaddr_in sadr;
	int rlen, rc, qid;
	int prand = getpid();
	struct nsnode *tns = nshead;
	in_addr_t ipno;
	int sent = 0;

	while(tns) {
		if (tns->resprec == 0) {
			qid = tns->queryno;
			rlen = dnsencode(who, buf, qid, authq?0:1, qtype, 1);
			bzero(&sadr, sizeof(sadr));
			sadr.sin_family = AF_INET;
			sadr.sin_port = htons(53);
			ipno = inet_addr(tns->nsip);
			if (ipno == INADDR_NONE) {
				upchuck("Bad IP number", tns->nsip);
				}
			sadr.sin_addr.s_addr = ipno;
			rc = sendto(dnsfd, buf, rlen, 0, (struct sockaddr *)&sadr, sizeof(sadr));
			if (rc == -1) barf("sendto");
			if (debug) fprintf(stderr, "send packet %d to %s\n", qid, tns->nsip);
			sent++;
			}
		tns = tns->next;
		}
	return sent;
	}

void dnsrec(unsigned char *buf, int blen, int rind, int nameoff, unsigned int rtype, unsigned int rclass, unsigned int rttl, unsigned int rsize, unsigned int recoff, void *edata) {
	char *aty[] = { "Query", "Answer", "Authority", "Additional", };
	char sbuf[256];
	struct nsnode *nsp = (struct nsnode *)edata;
	int rc;
	rc = dnsgetstr(buf, blen, nameoff, sbuf, sizeof(sbuf));
	if (debug) printf("%s", rc < 0 ? "ERROR" : sbuf);
	rc = dnsrrstr(buf, blen, rtype, rclass, rsize, recoff, sbuf, sizeof(sbuf));
	if (debug) printf(" - %s %s\n", aty[rind], rc ? "ERROR" : sbuf);
	if (rc == 0 && rind == 1) {
		struct rrnode *prr = findrr(sbuf);
		prr->bits |= nsp->npos;
		}
	}

void my_read() {
	int blen, rc, qid;
	char buf[1024];
	struct nsnode *nsp = nshead;
	blen = read(dnsfd, buf, sizeof(buf));
	if (blen == -1) barf("read");
	qid = dnsgetid(buf);
	rc = dnsgetrcode(buf);
	if (debug) fprintf(stderr, "got response %d %d (bytes %d)\n", qid, rc, blen);
	nsp = nshead;
	while(nsp) {
		if (nsp->queryno == qid) {
			if (debug) fprintf(stderr, "parsing response...\n");
			if (rc == 0) {
				dnsparse(buf, blen, dnsrec, (void*)nsp);
				nsp->resprec = 1;
				}
			else if (rc == 3) {
				nsp->resprec = -3;
				}
			else {
				nsp->resprec = -1;
				}
			break;
			}
		nsp = nsp->next;
		}
	}

void sloop() {
	int rc;
	fd_set readfds, writefds, exceptfds;
	struct timeval timeout;
	int tmdo = 1;
	int nsecs = 5;
	while(1) {
		if (tmdo) {
			if (nsecs-- == 0) return;
			if (sendqueries() == 0) return;
			}
		FD_ZERO(&readfds);
		FD_ZERO(&writefds);
		FD_ZERO(&exceptfds);
		FD_SET(dnsfd, &readfds);
		FD_SET(dnsfd, &exceptfds);
		timeout.tv_sec = 1;
		timeout.tv_usec = 0;
		tmdo = 0;
		if (debug) fprintf(stderr, "selecting...\n");
		rc = select(dnsfd+1, &readfds, &writefds, &exceptfds, &timeout);
		if (rc > 0) {
			if (debug) fprintf(stderr, "something...\n");
			if (FD_ISSET(dnsfd, &readfds)) {
				my_read();
				}
			else if (FD_ISSET(dnsfd, &exceptfds)) {
				puke("Exception on DNS socket");
				}
			}
		else if (rc < 0 && errno != EINTR) {
			barf("select()");
			}
		else {
			tmdo = 1;
			}
		}
	}
